### PR TITLE
MAINT:  better MemoryError message (#13225)

### LINKED ISSUE #13225: Indicate attempted allocation size when constructing an array gives a MemoryError
I've seen multiple questions now of the form "why does this give a MemoryError" when the reason is the user ended up allocating a bigger array than expected.

It would be helpful to include the shape and dtype in the error message.

### PR DESCRIPTION
**Pull request by bmakos**
I am really not sure if this is the best place to place this message and if it is the good call to suppress other debug info. 

Closes #13225.

### CODE REVIEW

**Review by eric-wieser — CHANGES_REQUESTED** [review]
This isn't a good place to put this message. It should go in `PyArray_New...` or something, right at the source. Search for `PyErr_NoMemory()` to find places to put it.

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -1072,7 +1072,17 @@ PyArray_NewFromDescr_int(PyTypeObject *subtype, PyArray_Descr *descr, int nd,
...
+            PyObject *errmsg = PyUString_FromString("Array shape:(");
+            for (i = 0; i < fa->nd-1; i++) {
+                PyUString_ConcatAndDel(&errmsg,
+                    PyUString_FromFormat("%ld,",fa->dimensions[i]));
+            }
+            PyUString_ConcatAndDel(&errmsg,
+                PyUString_FromFormat("%ld); Data type:",fa->dimensions[i]));
+            PyUString_ConcatAndDel(&errmsg,
+                PyUString_FromFormat("%s.",fa->descr->typeobj->tp_name));
+            PyErr_SetObject(PyExc_MemoryError, errmsg);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Unforunately you need to check every single function call here for a `NULL` return value, else you risk producing a segfault.

This is why I've been increasingly leaning towards doing the string formatting in python.

File: numpy/core/src/multiarray/ctors.c [hunk 2]
```diff
@@ -1072,17 +1072,19 @@ PyArray_NewFromDescr_int(PyTypeObject *subtype, PyArray_Descr *descr, int nd,
...
-            PyObject *errmsg = PyUString_FromString("Array shape:(");
-            for (i = 0; i < fa->nd-1; i++) {
-                PyUString_ConcatAndDel(&errmsg,
-                    PyUString_FromFormat("%ld,",fa->dimensions[i]));
+            PyObject *err_shape = PyArray_IntTupleFromIntp(fa->nd,fa->dimensions);
+            static PyObject *errmsg_formatter = NULL;
+			npy_cache_import("numpy.core._internal",
+                     "array_memory_errmsg_formatter",
+                     &errmsg_formatter);
+            if (errmsg_formatter != NULL) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: If `errmsg_formatter` returns `NULL` you should propagate it with `return NULL`

File: numpy/core/src/multiarray/ctors.c [hunk 3]
```diff
@@ -1072,17 +1072,19 @@ PyArray_NewFromDescr_int(PyTypeObject *subtype, PyArray_Descr *descr, int nd,
             data = npy_alloc_cache(nbytes);
         }
         if (data == NULL) {
-            PyObject *errmsg = PyUString_FromString("Array shape:(");
-            for (i = 0; i < fa->nd-1; i++) {
-                PyUString_ConcatAndDel(&errmsg,
-                    PyUString_FromFormat("%ld,",fa->dimensions[i]));
+            PyObject *err_shape = PyArray_IntTupleFromIntp(fa->nd,fa->dimensions);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Need to check if this returns `NULL` too

**Comment by eric-wieser** [comment]
Another way you could handle this is via the `_exception`s route used in `type_resolution.c`, which looks like:
```
    static PyObject *exc_type = NULL;
    exc_value;

    npy_cache_import(
        "numpy.core._exceptions", "_ArrayMemoryError",
        &exc_type);
    if (exc_type == NULL) {
        return NULL;
    }

    PyObject *shape = PyArray_IntTupleFromIntp(fa->nd,fa->dimensions);
    if (shape == NULL) {
        return NULL;
    }

    /* produce an error object */
    PyObject *exc_value = PyTuple_Pack(2, shape, descr);
    Py_DECREF(shape);
    if (exc_value == NULL){
        return NULL;
    }
    PyErr_SetObject(exc_type, exc_value);
    Py_DECREF(exc_value);
    return NULL;
```

**Comment by bmakos** [comment]
Who should solve the merge conflict? Should I fetch the new code and re-do my change?

**Comment by mattip** [comment]
@bmakos you should solve the conflicts either by merging master into your branch or (better) rebasing off updated master, and force-pushing to your branch. If you rebase, you can squash your commits to one. Be careful when squashing, if you "reach back too far" when choosing the first commit, you will end up rewriting history of commits that are not yours.

**Comment by bmakos** [comment]
I think I did the rebase and squash, but it took several attempts and re-cloning of my branch... 
`git log -p numpy/core/_exceptions.py` looks ok now, hope I did not mess up anything else.

File: numpy/core/_exceptions.py [hunk 4]
```diff
@@ -121,3 +121,17 @@ def __init__(self, axis, ndim=None, msg_prefix=None):
...
+class _ArrayMemoryError(MemoryError):
+    """ Thrown when an array cannot be allocated"""
+    def __init__(self, shape, dtype):
+        self.shape = shape
+        self.dtype = dtype
+
+    def __str__(self):
+        return "Attempted to allocate array with shape " + \
+        str(self.shape).rstrip(')').rstrip(',') + ") and data type " + \
+        str(self.dtype) + "."
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py]: I don't think you should remove the trailing comma from the `shape`.

Also, consider using `.format` here, as
```python
return "Unable to allocate array with shape {} and data type {}".format(self.shape, self.dtype)
```

**Review by eric-wieser — APPROVED** [review]
Looks good. I'll try testing this on my windows machine and see if I can make this message appear.

**Comment by eric-wieser** [comment]
Looks great:
```python
In [2]: np.zeros(10000000000000)
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-2-e424a0b4f3d7> in <module>()
----> 1 np.zeros(10000000000000)

MemoryError: Unable to allocate array with shape (10000000000000,) and data type float64
```
And just for curiosity
```
In [3]: import sys

In [4]: sys.last_value
Out[4]: numpy.core._exceptions.MemoryError((10000000000000,), dtype('float64'))
```

**Comment by eric-wieser** [comment]
@charris, do you think this would be a nice addition to the release notes? Something like

```
MemoryErrors caused by allocated overly large arrays are more descriptive
---
Often the cause of a MemoryError is incorrect broadcasting resulting in a very large and incorrect shape.
The message of the error now includes this shape to help diagnose its cause.
```

**Comment by bmakos** [comment]
@eric-wieser Thank you for walking me through this change, I imagine it took some patience. I struggled a bit with it, but on the way I learned a lot.

**Comment by charris** [comment]
> do you think this would be a nice addition to the release notes?

Sure, there is an "Improvements" section. The notes are more likely to be  "too few" than "too many".

**Comment by mattip** [comment]
Added release note. Will merge if build is OK

**Comment by mattip** [comment]
Thanks @bmakos

**Comment by opprash** [comment]
numpy.core._exceptions.MemoryError: Unable to allocate array with shape (9681, 9681) and data type float64
how to solve this problem？

**Comment by bmakos** [comment]
> numpy.core._exceptions.MemoryError: Unable to allocate array with shape (9681, 9681) and data type float64
> how to solve this problem？

It just means you ran out of memory. You could allocate a smaller array, or of a type that uses less space. You could also have more memory on your system. The solution really depends on what are you trying to do. Maybe post the code that lead to this error?

**Comment by seberg** [comment]
Oh, and I thought he meant that the full module of the `MemoryError` should not show up ;). That array does not seem to be awfully big, less than one GiB. Maybe creating many.

**Comment by opprash** [comment]
> > numpy.core._exceptions.MemoryError：无法分配带形状的数组（9681,9681）和数据类型float64 
> > 如何解决这个问题？
> 
> 它只是意味着你的内存不足。您可以分配较小的数组，或使用较少空间的类型。您还可以在系统上拥有更多内存。解决方案真的取决于你想做什么。也许发布导致此错误的代码？
> 
> 这个帖子也可能不是讨论这个问题的最佳场所。

Traceback (most recent call last):
  File "C:/Users/ruanlu/PycharmProjects/part_nlp/dealText/tf_idf.py", line 311, in <module>
    tfidf_mat = get_all_vector(path, stop_words)
  File "C:/Users/ruanlu/PycharmProjects/part_nlp/dealText/tf_idf.py", line 108, in get_all_vector
    idf = np.diag(idf)
  File "<__array_function__ internals>", line 6, in diag
  File "C:\Users\ruanlu\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\lib\twodim_base.py", line 275, in diag
    res = zeros((n, n), v.dtype)
numpy.core._exceptions.MemoryError: Unable to allocate array with shape (11219, 11219) and data type float64
The above is my specific error message, here I use a length array 11219, type array float idf of type float64, call np.diag (idf) to construct the error of the two-dimensional vector time report

**Comment by eric-wieser** [comment]
@seberg: I think we should maybe lock this thread, its not a good place to handle this type of report.

**Comment by opprash** [comment]
> @seberg: I think we should maybe lock this thread, its not a good place to handle this type of report.
Sorry, I should not discuss it in this place.

**Comment by seberg** [comment]
Yeah, OK, lets lock it down.

#### If you came to this issue because saw this particular error message:

You tried to create an array and this failed because there was not enough Memory available. There are two reasons for this:
  1. The array you are creating is too large, likely by accident (inspect the shape and data type provided by the error message)
  2. Other parts of your program already use a lot of memory/RAM.

In most cases, you probably should find out what is using the memory and avoid it.

**Pull request merged**